plugins/inputs/snmp_trap/snmp_trap.go

```diff
@@ -0,0 +1,251 @@
+	return nil
+}
+
+func (s *SnmpTrap) Start(acc telegraf.Accumulator) error {
+	s.acc = acc
+	s.listener = gosnmp.NewTrapListener()
+	s.listener.OnNewTrap = makeTrapHandler(s)
+	s.listener.Params = gosnmp.Default
+
+	// wrap the handler, used in unit tests
+	if nil != s.makeHandlerWrapper {
+		s.listener.OnNewTrap = s.makeHandlerWrapper(s.listener.OnNewTrap)
+	}
+
+	split := strings.SplitN(s.ServiceAddress, "://", 2)
+	if len(split) != 2 {
+		return fmt.Errorf("invalid service address: %s", s.ServiceAddress)
+	}
+
+	protocol := split[0]
+	addr := split[1]
+
+	// gosnmp.TrapListener currentl supports udp only.  For forward
+	// compatibility, require udp in the service address
+	if protocol != "udp" {
+		return fmt.Errorf("unknown protocol '%s' in '%s'", protocol, s.ServiceAddress)
+	}
+
+	// If (*TrapListener).Listen immediately returns an error we need
+	// to return it from this function.  Use a channel to get it here
+	// from the goroutine.  Buffer one in case Listen returns after
+	// Listening but before our Close is called.
+	s.errCh = make(chan error, 1)
+	go func() {
+		s.errCh <- s.listener.Listen(addr)
+	}()
+
+	select {
+	case <-s.listener.Listening():
+		s.Log.Infof("Listening on %s", s.ServiceAddress)
+	case err := <-s.errCh:
+		return err
+	}
+
+	return nil
+}
+
+func (s *SnmpTrap) Stop() {
+	s.listener.Close()
+	_ = <-s.errCh
+}
+
+func makeTrapHandler(s *SnmpTrap) handler {
+	return func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr) {
+		tm := s.timeFunc()
+		fields := map[string]interface{}{}
+		tags := map[string]string{}
+
+		tags["version"] = packet.Version.String()
+		tags["source"] = addr.IP.String()
+
+		for _, v := range packet.Variables {
+			// Use system mibs to resolve oids.  Don't fall back to
+			// numeric oid because it's not useful enough to the end
+			// user and can be difficult to translate or remove from
+			// the database later.
+
+			var value interface{}
+
+			// todo: format the pdu value based on its snmp type and
+			// the mib's textual convention.  The snmp input plugin
+			// only handles textual convention for ip and mac
+			// addresses
+
+			switch v.Type {
+			case gosnmp.ObjectIdentifier:
+				val, ok := v.Value.(string)
+				var e mibEntry
+				var err error
+				if !ok {
+					s.Log.Errorf("Error getting value OID: %v", err)
+					return
+				}
+
+				e, err = s.lookup(val)
+				if nil != err {
+					s.Log.Errorf("Error resolving value OID: %v", err)
+					return
+				}
+
+				value = e.oidText
+
+				// 1.3.6.1.6.3.1.1.4.1.0 is SNMPv2-MIB::snmpTrapOID.0.
+				// If v.Name is this oid, set a tag of the trap name.
+				if v.Name == ".1.3.6.1.6.3.1.1.4.1.0" {
+					tags["oid"] = val
+					tags["name"] = e.oidText
+					tags["mib"] = e.mibName
+					continue
+				}
+			default:
+				value = v.Value
+			}
+
+			e, err := s.lookup(v.Name)
+			if nil != err {
+				s.Log.Errorf("Error resolving OID: %v", err)
+				return
+			}
+
+			name := e.oidText
+
+			fields[name] = value
```

Comment:
What are the possible types?  Do we need to do any conversions?

Reply:
I suspect we'll need to do more work converting and formatting.  To do that I need some real world traps to test with.  I was hoping we could get this out and ask people watching the feature request for test cases.